transcom/mymove
# INT B-21702

## [Agility ticket](https://www13.v1host.com/USTRANSCOM38/assetdetail.v1?number=B-21702)

**NOTE**
This work was done off of `integrationTesting` since the ordering/pricing/port work is already in here.

**SECOND NOTE**
Please test these carefully & thoroughly

## Summary

When we order basic iHHG service items, we also want the ability to see the estimated price when we have the estimated weight on the shipment. In order to do this, the UI has to allow the estimated price section show for iHHG service items AND we have to populate the `pricing_estimate` data any time that pertinent data is provided and/or changed.

This PR contains a buttload of file changes because I changed the `db-truncate` to stop deleting static data we use for pricing & other additional params. This required the creation of a couple factories, and many many many changes in test data because now we are using data pulled from the db instead of data created within tests & then saved to the test db.

In a nutshell, we want to calculate estimated pricing & display that estimated pricing when the following things happen:
- shipments are updated with relevant values (that differ from the previous shipment data)
- shipment address updates are approved (address change == distance change)
- shipments that are created by Prime when they provide the estimated weight

Some stickler details we need to ensure we have:
- the `poe_location` or `pod_location` for `POEFSC` or `PODFSC` basic service items
- `shipment.prime_estimated_weight`
- `addresses.us_post_region_cities_id`

Without the above details, pricing ain't gonna happen.

Some additional things happening in this PR with migrations:
- updating the `get_zip_code` function to not return an error, this was causing unnecessary errors when we didn't have the port location stuff yet
- updating the `get_rate_area_id` function to take in the `contract_id` - this is useful when there are multiple contracts
- adding a `get_port_location_info_for_shipment` function that will return the port ZIP associated with the shipment as well as if it is a `POEFSC` or `PODFSC`
- updating the pricing proc with updated parameters since we changed the parameters of `get_rate_area_id` as well as taking in a `mileage` value that we are getting from DTOD prior to the pricing func being called - this is a business requirement and AC on the feature

### How to test

1. You need to first create an `iHHG` shipment (don't approve yet) - do that however you wish, but please keep in mind we do not have total coverage on CONUS -> AK OCONUS ZIP codes (I have had good luck with ZIPs of major bases and the `998` ZIP
2. Submit the move details as SC
3. Approve the shipment as TOO
4. Verify that the four basic service items have been created and they don't have a `pricing_estimate` value on any of the service items
5. Update the estimated weight as the Prime and other required details
6. You should STILL see a null `pricing_estimate` value on the service items - this is because we need the mileage from/to the port and since we don't have a `poe_location_id` or `pod_location_id` on the `POEFSC`/`PODFSC` service item
7. The Prime will need to update that value, but you're gonna need two things: the `eTag` for the fuel surcharge service item (Prime v3 `getMoveTaskOrder`) & the service item `id` - [swagger UI makes it easy](http://primelocal:3000/swagger-ui/prime_v3.html#/moveTaskOrder/getMoveTaskOrder)
8. Now you will need to update the port of the service item by using [prime v1 updateServiceItem](http://officelocal:3000/swagger-ui/prime.html#/mtoServiceItem/updateMTOServiceItem) - make sure you got that `eTag`
9. Payload can look like this (using your respective values):
```
{   
   "id": "4cc44e88-231f-4f49-8180-4d7012be17b3",   
   "modelType": "UpdateMTOServiceItemInternationalPortFSC",   
   "portCode": "SEA",   
   "reServiceCode": "POEFSC" 
}
```
10. You should now see updates to the `pricing_estimate` values on all the basic service items
11. You can now test doing shipment address updates and see changing the respective address values (dest address changes should only impact shipments with `PODFSC` - which are `OCONUS` -> `CONUS` shipments
12. You can also test changing the requested pickup date to be in a different contract period - this should also update the pricing. This can be done as TOO.

- ignore the negative values here, the test data we have in `ghc_diesel_fuel_prices` is lower than $2.50 so a lot of the times I get a negative value for that. You could update the last row in that table to be something greater (say `400,000` instead of `243,300` and that value will be positive. You can also run `make tasks_save_ghc_fuel_price_data` to try and get updated values imported into your dev db.
![Screenshot 2024-12-20 at 10 58 28 AM](https://github.com/user-attachments/assets/337fad31-63ac-4878-992f-d2c365daa066)
